File: python/meetup/meetup.py
from calendar import Calendar
from datetime import date

WEEK_DAYS = ['Monday', 'Tuesday', 'Wednesday',
             'Thursday', 'Friday', 'Saturday', 'Sunday', ]


def meetup(year, month, week, day_of_week):
    days = [(d, w) for d, w in Calendar().itermonthdays2(year, month)
            if d != 0 and WEEK_DAYS[w] == day_of_week]

    if week == 'teenth':
        day = next(d for d, _ in days if 13 <= d <= 19)
    elif week == 'last':
        day = days[-1][0]
    else:
        try:
            day = days[int(week[0]) - 1][0]
        except IndexError:
            raise MeetupDayException(r".+")

    return date(year, month, day)


class MeetupDayException(Exception):
    pass

File: haskell/meetup/src/Meetup.hs
module Meetup
  ( Weekday(..)
  , Schedule(..)
  , meetupDay
  ) where

import           Data.Time.Calendar             ( Day
                                                , dayOfWeek
                                                , fromGregorian
                                                )

data Weekday = Monday
             | Tuesday
             | Wednesday
             | Thursday
             | Friday
             | Saturday
             | Sunday
             deriving (Enum, Eq)

data Schedule = First
              | Second
              | Third
              | Fourth
              | Last
              | Teenth
              deriving (Enum, Eq)

meetupDay :: Schedule -> Weekday -> Integer -> Int -> Day
meetupDay s w y m | s == Teenth = head days''
                  | s == Last   = last days''
                  | otherwise   = days'' !! fromEnum s
 where
  days   = if s /= Teenth then [1 .. 31] else [13 .. 19]
  days'  = map (fromGregorian y m) days
  days'' = filter ((== w) . toEnum . pred . fromEnum . dayOfWeek) days'
